### DISCUSSION TITLE
❓ [QUESTION] How exactly does the FullyConnectedTensorProduct work?

### DISCUSSION THREAD

**Discussion by iorin0217** [e3nn/e3nn#288] [Q&A] [answered]
I'm confusing about learnable weights implemented in `FullyConnectedTensorProduct`.
For example
```
tp = FullyConnectedTensorProduct(irreps_in1="1x0e",irreps_in2="2x0e",irreps_out="1x0e")
w = list(tp.weight_views())[0]
x = o3.Irreps("1x0e").randn(-1)
y = o3.Irreps("2x0e").randn(-1)
```
The values are
```
w: tensor([[[-0.0048],
         [-2.1463]]], grad_fn=<ViewBackward>)
x: tensor([0.7021])
y: tensor([-0.9132,  0.8710])
tp(x,y): tensor([-0.9259], grad_fn=<ViewBackward>)
w[0][0]*x[0]*y[0]+w[0][1]*x[0]*y[1]: tensor([-1.3094], grad_fn=<AddBackward0>)
```
Why `tp(x,y)` is not equals to `w[0][0]*x[0]*y[0]+w[0][1]*x[0]*y[1] `?

**Comment by mariogeiger** [comment]
Because there is a normalization constant. Here it will be 1/sqrt(2). 

Normalization scheme:
1. inputs have mean(x**2)=1
2. each output is mean 0 and variance 1 (expectation over the weights distribution)
3. each weight is sampled from a normal gaussian (mean 0 and variance 1)

=> multiplicative constants are added to enforce point 2

**Comment by iorin0217** [comment]
Thank you for your polite response.
I also understand why there are input options such as `in1_var`.

By the way, I have another question about Clebsch–Gordan coefficients.
1. I have compared `o3.wigner_3j` with the `Wigner3j` implemented in `Sympy` ([https://docs.sympy.org/latest/modules/physics/quantum/cg.html](https://docs.sympy.org/latest/modules/physics/quantum/cg.html)), and these values do not match. What's going on here?

```python
from sympy.physics.quantum.cg import Wigner3j
o3.wigner_3j(1,2,3)
'''
tensor([[[-0.0000, -0.0000, -0.0000, -0.0000, -0.0690, -0.0000, -0.2673],
         [-0.0000, -0.0000, -0.0000, -0.1690, -0.0000, -0.2182, -0.0000],
         [-0.0000, -0.0000,  0.2390, -0.0000, -0.0000, -0.0000, -0.0000],
         [-0.0000,  0.2182, -0.0000, -0.0000, -0.0000, -0.0000, -0.0000],
         [ 0.2673, -0.0000,  0.0690, -0.0000, -0.0000, -0.0000, -0.0000]],

        [[-0.0000,  0.2182, -0.0000, -0.0000, -0.0000, -0.0000, -0.0000],
         [-0.0000, -0.0000,  0.2760, -0.0000, -0.0000, -0.0000, -0.0000],
         [-0.0000, -0.0000, -0.0000,  0.2928, -0.0000, -0.0000, -0.0000],
         [-0.0000, -0.0000, -0.0000, -0.0000,  0.2760, -0.0000, -0.0000],
         [-0.0000, -0.0000, -0.0000, -0.0000, -0.0000,  0.2182, -0.0000]],

        [[ 0.2673, -0.0000, -0.0690, -0.0000, -0.0000, -0.0000, -0.0000],
         [-0.0000,  0.2182, -0.0000, -0.0000, -0.0000, -0.0000, -0.0000],
         [-0.0000, -0.0000, -0.0000, -0.0000,  0.2390, -0.0000, -0.0000],
         [-0.0000, -0.0000, -0.0000, -0.1690, -0.0000,  0.2182, -0.0000],
         [-0.0000, -0.0000, -0.0000, -0.0000, -0.0690, -0.0000,  0.2673]]])
'''
w3j = Wigner3j(1,0,2,0,3,0)
w3j.doit() # -0.2927700218845599
w3j = Wigner3j(1,-1,2,-1,3,2)
w3j.doit() # -0.3086066999241838
w3j = Wigner3j(1,0,2,-2,3,2)
w3j.doit() # -0.21821789023599236
w3j = Wigner3j(1,-1,2,-2,3,3)
w3j.doit() # 0.37796447300922725
```

2. Does the `m` index of `Y_{lm}` correspond to the index of `o3.wigner_3j`? For example, `o3.wigner_3j(1,2,3)[0,2,6]` means `(l,m) = (1,-1), (2,0), (3,3)`?

3. Sorry if I'm wrong, why using Wigner 3j instead of Clebsch–Gordan coefficients to calculate Tensor Products?

**Comment by mariogeiger** [comment]
We don't use the same irreps as sympy. there is a change of basis (for each l) between our irreps and sympy irreps. They have complex irreps and we have a real one. 

Wigner 3j and Clebsh Gordan are the same up to a constant (constant that depends on the l's).

**Reply by Glinttsd** [reply]
Hi, I'm curious about the explicit mathematical expression of `o3.wigner_3j` because I've noticed inconsistencies between the results from my implementation according to [wiki](https://en.wikipedia.org/wiki/3-j_symbol) and the results from `o3.wigner_3j`.

I would greatly appreciate any assistance. Thank you!

**Reply by mitkotak** [reply]
This may or may not help but here's a sympy version https://github.com/atomicarchitects/BasisLib/blob/123d9d8157af405252198ab259805e8f2eac50dd/BasisLib/so3/symbolic.py#L66-L113.

**Reply by Glinttsd** [reply]
Thanks a lot!

**Comment by blondegeek** [comment]
FullyConnectedTensorProduct has weights on independent paths.
FullTensorProduct is more directly comparable to o3.wigner_3j but
additionally there are normalization considerations. Please see the docs
https://docs.e3nn.org/en/stable/api/o3/o3_tp.html for more details.

Best,
Tess

**Reply by Glinttsd** [reply]
Hi Tess,

I checked the source code of [wigner_3j](https://github.com/e3nn/e3nn/blob/ac3528f7fb5fe1a8838f1df087d2eefe60d91ea8/e3nn/o3/_wigner.py#L100), which indicates that the wigner_3j matrix is generated from the SU(2) Clebsch Gordan coefficient followed by some matrix multiplications to change the basis ([ref](https://github.com/e3nn/e3nn/blob/ac3528f7fb5fe1a8838f1df087d2eefe60d91ea8/e3nn/o3/_wigner.py#L149)).

My question is that: Is there a uniformed (or simplified) expression of SO(3) Clebsch Gordan coefficient? 

I asked because I've noticed the inconsistencies between the code and the expression from [wiki](https://en.wikipedia.org/wiki/Table_of_Clebsch%E2%80%93Gordan_coefficients#_j1_=_1,_j2_=_1), where a simple formula is given.

Any help will be greatful!

Best,
Shidi

**Comment by blondegeek** [comment]
This probably won't fully answer your question, but an important thing to
note is that there is no unique set of Clebsch-Gordan coefficients and
there are in fact an infinite number of valid changes of basis. e3nn uses a
particular convention such that our L=1 irreps transform as (x, y, z) and
we also use specific normalization factors in our tensor product that is
favorable for training neural networks. So the important thing is not
whether it matches a specific convention but rather that the chosen basis
is consistent and are indeed irreducible representations of SO(3) (because
it's easier to parameterize equivariant weights for irreps than other
representations).

As for is there simplified expression for Clebsch-Gordan coefficients -- I
think the code here is fairly readable:
https://github.com/e3nn/e3nn/blob/0.5.0/e3nn/o3/_wigner.py#L55-L92

In my symmetry for ML class, I teach how to do this numerically rather than
using closed form solutions which I think is better for building intuition
-- hopefully I will feel that the material is ready to make public in the
next year or so.

**Reply by Glinttsd** [reply]
Thank you for your patience! I am looking forward to your class next year.

I have one more question that is related to the project I am working on:

Do all sets (kinds) of CG coefficients hold the properties of symmetry (i.e., C^{(l3,m3)}\_{(l1,m1)(l2,m2)} = (-1)^{l1+l2-l3} * C^{(l3,-m3)}\_{(l1,-m1)(l2,-m2)} ) and orthogonality?

I've noticed that the output matrix of `o3.wigner_3j` ([ref](https://github.com/e3nn/e3nn/blob/ac3528f7fb5fe1a8838f1df087d2eefe60d91ea8/e3nn/o3/_wigner.py#L100)) is **not** symmetry, but the output matrix of `_su2_clebsch_gordan` ([ref](https://github.com/e3nn/e3nn/blob/ac3528f7fb5fe1a8838f1df087d2eefe60d91ea8/e3nn/o3/_wigner.py#L153)) is symmetry, which means that the changes of basis ([ref](https://github.com/e3nn/e3nn/blob/ac3528f7fb5fe1a8838f1df087d2eefe60d91ea8/e3nn/o3/_wigner.py#L150-L152)) destroys the property of symmetry.

Much appreciated!

Best,
Shidi

**Reply by mariogeiger** [reply]
Indeed we trade off that nice symmetry for a real basis

**Reply by Glinttsd** [reply]
Thanks a lot. That answers my question!

**Comment by blondegeek** [comment, marked as answer]
Yes on orthogonality not necessarily on that -1 factor. I’m not positive,
but that -1 factor might be dependent on whether it’s a real or complex
basis. In e3nn we choose a real basis for our irreps for so(3).

**Reply by Glinttsd** [reply]
Thanks a lot. That answers my question!
